Implement gradient outlines

adroitwhiz opened the pull request:
## Depends on https://github.com/LLK/paper.js/pull/39, https://github.com/LLK/paper.js/pull/40, https://github.com/LLK/scratch-paint/pull/1166

### Resolves

Resolves #540
Diff: https://github.com/adroitwhiz/scratch-paint/compare/stroke-gradient-part-1...adroitwhiz:stroke-gradient?expand=1

### Proposed Changes

This PR adds support for (and contains a bunch of supporting changes for) setting outlines' colors to gradients in select mode, fill mode, and shape tools:
![Peek 2020-04-15 12-40](https://user-images.githubusercontent.com/25993062/79363851-73618380-7f16-11ea-89a8-6a12c527bbc2.gif)

The supporting changes include:
- Combining "fill color", "fill color 2", and "gradient type" reducers into a single "swatch" reducer. A "swatch" (not sure what else to call it) consists of a primary color, a secondary color, and a gradient type (solid, horizontal, vertical, or radial).
  - I've further abstracted the "swatch" reducer into a swatch-reducer-generator function that generates swatch reducers given the names of actions it should respond to. This allows logic to be shared between fill and stroke reducers while still allowing each to have its own unique action names.
- Modifying all helper functions in `style-path.js` to work on both fill and stroke.
- Abstracting `FillColorPickerComponent` and `StrokeColorPickerComponent` into a `ColorPickerComponent`.
- Abstracting the `FillColorPicker` and `StrokeColorPicker` containers into a `ColorPicker` container component generator.
  - As with the swatch reducer generator, this allows for multiple color picker containers that share the same logic, but are mapped to different Redux state.

~~It deliberately does *not* yet add support for gradient strokes in fill mode-- there's some behavior there that would best be ironed out at some point in the future.~~
Per design meeting we agreed to include gradient outlines in select mode, gradients in shape tools, and gradient outlines in the fill tool in one release

### Reason for Changes

This has been a commonly-requested feature ever since 3.0 was first released to the public.

The supporting changes should also make it easier to implement things like #1003.

### Test Coverage

Existing fill/outline color reducer tests have been merged together into one big color reducer test

adroitwhiz committed:
Simplify fill-color-2 reducer logic

why my brain hurt???

adroitwhiz committed:
Combine fill-color and fill-color-2 reducers

adroitwhiz committed:
Abstract color reducer into reducer-generator

adroitwhiz committed:
color -> style, move selection gradient into it

A "style" refers to something that can fill/stroke a shape.
Currently that's either a solid color or a gradient of some orientation.
The selection gradient type reducer has been removed and folded into the
"fill style" reducer.

adroitwhiz committed:
Make style-path helpers generic over fill & stroke

adroitwhiz committed:
Wire up stroke gradient controls

* Abstract FillColorIndicator and StrokeColorIndicator to ColorIndicator
* Replace stroke color reducer with stroke style reducer
* Add color style proptype
* Clear stroke gradient in line mode

adroitwhiz committed:
Update color reducer tests

fsih commented:
This is awesome!
Adding @carljbowman for design feedback

adroitwhiz committed:
Fix setting of stroke colors in bitmap mode

adroitwhiz committed:
Add gradients to vector shape tools

adroitwhiz commented:
![image](https://user-images.githubusercontent.com/25993062/82849375-a445b880-9ec5-11ea-933a-4a5788383dcc.png)

These aren't showing up properly on the stage

adroitwhiz committed:
Add hit result to hovered item data

adroitwhiz committed:
Make fill tool work for outlines

adroitwhiz committed:
Don't set color in bitmap shape onSelectionChanged

This *should* be safe because OvalTool and RectTool's wrapper "mode"
components, the only places that call onSelectionChanged, also update
the tools' color, and the color state reducers will always set the color
every time the selection changes, meaning it'll be updated anyway.

adroitwhiz committed:
Add gradients to bitmap shape tools

fsih commented on src/containers/bit-brush-mode.jsx:
```diff
@@ -5,11 +5,10 @@ import bindAll from 'lodash.bindall';
 import {MIXED} from '../helper/style-path';
 
-import {changeFillColor, DEFAULT_COLOR} from '../reducers/fill-color';
+import {changeFillColor, clearFillGradient, DEFAULT_COLOR} from '../reducers/fill-swatch';
```
Since swatches are going to be a separate thing, we probably want to choose a different word...

fsih replied:
fill-gradient ??

fsih replied:
fill-color-or-gradient?

adroitwhiz replied:
Note to self: fill-style and stroke-style

fsih commented on src/reducers/fill-color-2.js:
```diff
@@ -24,13 +24,12 @@ const reducer = function (state, action) {
-            return state;
+        // Gradient type may be solid when multiple gradient types are selected.
+        // In this case, changing the first color should not change the second color.
+        if (colors.gradientType !== GradientTypes.SOLID || colors.fillColor2 === MIXED) {
```
I think we figured out that the getColorsFromSelection function returns GradientType = SOLID, fill colors 1 and 2 = MIXED to indicate that the gradient type isn't actually solid, you've selected a bunch of things with different gradient types. (Just clarifying this for myself)

fsih commented on src/helper/tools/text-tool.js:
```diff
@@ -241,7 +241,8 @@ class TextTool extends paper.Tool {
                 font: this.font,
                 fontSize: 40,
-                fillColor: this.colorState.fillColor,
+                // TODO: style using gradient?
```
Might be worth filing a feature request issue and linking it here

fsih commented on src/lib/make-color-reducer.js:
```diff
@@ -0,0 +1,62 @@
+};
+
+const makeColorReducer = ({
+    changePrimaryColorAction,
```
Documenting these params would be helpful

fsih commented on src/helper/style-path.js:
```diff
@@ -112,40 +120,51 @@ const applyFillColorToSelection = function (colorString, colorIndex, isSolidGrad
+            continue;
+        }
+
+        // TODO: the JS engine may not approve of these dynamic property access shenanigans
```
Maybe could clarify that it means the JS performance would be better if we referred to properties by fixed strings

adroitwhiz committed:
Remove unnecessary bitmapMode params

fsih commented on src/helper/style-path.js:
```diff
@@ -190,51 +210,58 @@ const swapColorsInSelection = function (bitmapMode, textEditTargetId) {
  * Called when setting gradient type
  * @param {GradientType} gradientType gradient type
  * @param {?boolean} bitmapMode True if the fill color is being set in bitmap mode
+ * @param {boolean} applyToStroke True if changing the selection's stroke, false if changing its fill.
```
?boolean

fsih commented on src/helper/style-path.js:
```diff
@@ -245,73 +272,57 @@ const applyGradientTypeToSelection = function (gradientType, bitmapMode, textEdi
+        case GradientTypes.HORIZONTAL: {
+            const hasHorizontalGradient = hasGradient &&
+            !itemColor.gradient.radial &&
+            Math.abs(itemColor.origin.y - itemColor.destination.y) < 1e-8;
```
Indent and maybe newline after =

fsih commented on src/helper/style-path.js:
```diff
@@ -339,6 +350,31 @@ const applyStrokeWidthToSelection = function (value, textEditTargetId) {
+const _colorStateFromGradient = gradient => {
+    const colorState = {};
+    // Scratch only recognizes 2 color gradients
+    if (gradient.stops.length === 2) {
```
Maybe if gradient.stops.length < 2 there should be some kind of warning?

fsih commented on src/helper/style-path.js:
```diff
@@ -430,42 +477,62 @@ const getColorsFromSelection = function (selectedItems, bitmapMode) {
             }
         }
     }
     // Convert selection gradient type from horizontal to vertical if first item is exactly vertical
```
"This is because we assume all horizontal or vertical gradients are horizontal when comparing gradient types."

fsih reviewed (commented):
Let's separate everything up to e89bed9f as a "refactor" pr
(caveat: we need to have getColorsFromSelection return mixed for all stroke gradients until the rest is hooked in)

fsih commented on src/containers/color-indicator.jsx:
```diff
@@ -0,0 +1,142 @@
+            return (
+                <ColorIndicatorComponent
+                    {...this.props}
+                    label={this.props.intl.formatMessage(messages.label)}
```
Could we pass in messages.label instead of messages?
Otherwise, I think we should document that label is expected to be defined on the messages passed in

fsih commented on test/unit/color-reducer.test.js:
```diff
@@ -0,0 +1,86 @@
+            .toEqual(MIXED);
+    });
+
+    test('invalidchangeColor', () => {
```
caps invalidChangeColor

fsih commented on src/containers/text-mode.jsx:
```diff
@@ -82,8 +82,9 @@ class TextMode extends React.Component {
         // If exactly one of fill or stroke color is set, set the other one to transparent.
         // This way the tool won't draw an invisible state, or be unclear about what will be drawn.
-        const {strokeColor, strokeWidth} = nextProps.colorState;
+        const {strokeWidth} = this.props.colorState;
```
This looks like it should be nextProps

fsih replied:
I filed https://github.com/LLK/scratch-paint/issues/1141 to look into why nextProps is here in the first place

fsih commented on src/helper/tools/fill-tool.js:
```diff
@@ -31,6 +31,8 @@ class FillTool extends paper.Tool {
         // The path that's being hovered over.
         this.fillItem = null;
+        // The shape property that we're applying the color to (either fill or stroke).
+        this.fillType = null;
```
fillShapeProperty?

fsih commented on src/helper/tools/fill-tool.js:
```diff
@@ -43,21 +45,30 @@ class FillTool extends paper.Tool {
+            // prevent ourselves from hitting an outline when we make it transparent via the fill preview, causing it to
+            // flicker back and forth between transparent/its previous color as we hit it, then stop hitting it, etc.
+            // If the color *is* visible, then don't hit "invisible" outlines, since this would add visible outlines to
+            // non-outlined shapes when you hovered over where their outlines would be.
```
In Scratch 2.0, you can use the fill tool to add strokes to paths that don't have strokes ("invisible strokes"), and the stroke width would be set to some arbitrary relatively thick width. We can reproduce this by hitting curves and not strokes when the stroke is invisible, and overwriting whatever stroke width was on the invisible stroke before. If we want to change the behavior from 2.0, let's remember to ask about it in the paint huddle

fsih replied:
We conferred with design and decided not to "fill" invisible outlines for now

fsih commented on src/helper/style-path.js:
```diff
@@ -507,7 +507,10 @@ const getColorsFromSelection = function (selectedItems, bitmapMode) {
 
         if (selectionFillGradientType !== GradientTypes.SOLID) {
-            const direction = firstItem.fillColor.destination.subtract(firstItem.fillColor.origin);
+            // Stroke color is fill color in bitmap
```
"if fill color is missing"

fsih commented on src/containers/color-indicator.jsx:
```diff
@@ -39,30 +39,32 @@ const makeColorIndicator = (messages, isStroke) => {
-                isBitmap(this.props.format),
-                isStroke,
+                formatIsBitmap,
+                isStroke || (formatIsBitmap && !this.props.fillBitmapShapes),
```
comment here

fsih commented on src/containers/stroke-color-indicator.jsx:
```diff
@@ -37,7 +38,9 @@ const mapStateToProps = state => ({
-        state.scratchPaint.mode === Modes.OVAL,
+        state.scratchPaint.mode === Modes.OVAL ||
+        state.scratchPaint.mode === Modes.BIT_RECT ||
+        state.scratchPaint.mode === Modes.BIT_OVAL,
```
Maybe we should make a group for these in Modes.js, or a checker function?

fsih commented on src/helper/bitmap.js:
```diff
@@ -275,8 +275,23 @@ const drawEllipse = function (options, context) {
+    const isGradient = context.fillStyle instanceof CanvasGradient;
+
+    // If drawing a gradient, we need to draw the shape onto a temporary canvas, then draw the gradient atop that canvas
+    // only where the shape appears. Some points get double-filled which is a problem if the gradient fades to
```
Clarify that this is an issue with the drawShearedEllipse function that could be a problem if we become able to draw semitransparent ellipses. Maybe link to an issue?

fsih commented on src/helper/bitmap.js:
```diff
@@ -773,6 +823,57 @@ const commitSelectionToBitmap = function (selection, bitmap) {
+        let canvasGradient;
+        const {origin, destination} = color;
+        if (color.gradient.radial) {
+            // https://github.com/paperjs/paper.js/blob/b081fd72c72cd61331313c3961edb48f3dfaffbd/src/style/Color.js#L926-L935
```
"Adapted from"

adroitwhiz committed:
Add GradientToolsModes for shouldShowGradientTools

This means that gradient tools will also be enabled for the stroke color
indicator even in fill and bitmap modes, but that's okay because the
stroke color indicator will be disabled or hidden in those modes anyway.

fsih reviewed (commented):
I still think it might be a good idea to separate out the refactors to break up this large PR

adroitwhiz committed:
Make styleShape more flexible

Now, you can pass null in for a color instead of
{primary: null, secondary: null, gradientType: GradientTypes.SOLID}
and it'll still clear the color. Passing strokeWidth is also
optional now.

adroitwhiz committed:
Fix fill color reducer test

adroitwhiz commented:
This is now part 2 of 2 PRs-- part 1 is https://github.com/LLK/scratch-paint/pull/1166

fsih commented:
Bugs from bug hunt to fix:
- if you click the gradient button twice, the gradient disappears
- changing the first color of the gradient to transparent zeros out the stroke width
- brightness threshold for changing the second gradient color should be merged into the gradient out lines branch

adroitwhiz committed:
Fix changing gradient strokes to/from null

Hopefully the comments help; the logic is kinda convoluted here

adroitwhiz committed:
Use createGradientObject in applyGradientTypeToSelection

adroitwhiz committed:
Enforce minimum gradient size

This fixes the bug where percertly horizontal gradients on perfectly
vertical lines and vice versa would not be rendered at all

adroitwhiz committed:
Gradients in line mode

The UX here isn't the best thing in the world but I think having the
functionality is important judging from the playtest

fsih reviewed (commented):
It's looking really good!
I found that if I type 0 in the stroke width field, or press the down arrow key until it hits 0, the outline color doesn't get set to transparent, but maybe that's pre-existing

fsih commented:
Something strange
![bitmap color](https://user-images.githubusercontent.com/2855464/88584703-0c2c9280-d020-11ea-846b-0012bda2c21d.gif)

fsih commented:
Design meeting decided we should have the minimum gradient width equal the stroke width, to fix the instantaneous color change on vertical/horizontal lines

adroitwhiz committed:
Fix gradient outlines with 0 width in select mode

Previously, if you selected a shape with a gradient outline that had a
width of 0, its outline color wouldn't be null when it should have been.

adroitwhiz committed:
Fix zero-width-outline logic

- Set the stroke color to "null" when the width is set to 0
- Properly set the stroke color state when the width is increased from 0

adroitwhiz commented:
@fsih The latest 4 commits are new, and should fix issues relating to changing the outline width to 0.

Did you run into "something strange" in a local build or [my build](adroitwhiz.github.io/scratch-gui/gradient-outlines/)? It looks like you ran into the bug that https://github.com/LLK/paper.js/pull/40 fixes, so you may need to `npm link` paper.js, pull in that PR, and rebuild it.

fsih commented:
@adroitwhiz Yup, it was a local build. You're right, that's probably what it was!

fsih reviewed (approved):
Awesome! Let's do it!!

It looks like these issues are fixed as well:
https://github.com/LLK/scratch-paint/issues/738
https://github.com/LLK/scratch-paint/issues/697

fsih reviewed (dismissed):
ignore this comment

adroitwhiz committed:
Enforce minimum gradient width for fill tool

adroitwhiz commented:
@fsih I noticed one more thing while testing this out again-- https://github.com/LLK/scratch-paint/pull/1004/commits/379599905aa61df751d2dbd90bb5165a14f2ea6b makes it so that the fill tool now enforces the same minimum gradient size as the line tool does for perfectly horizontal/vertical lines

adroitwhiz committed:
Also enforce minimum gradient size in select mode

Oops

fsih reviewed (approved):
Good catches
